lib/getWorkflow.js

```diff
@@ -0,0 +1,119 @@
+    });
+
+    return nodes;
+};
+
+/**
+ * Get all the edges of the directed graph from a legacy workflow config
+ * @method calculateLegacyEdges
+ * @param  {Array}             workflow  List of all jobs in the workflow excluding "main"
+ * @return {Array}                       List of edge objects { src, dest }
+ */
+const calculateLegacyEdges = (workflow) => {
+    // In legacy-mode "main" is always required to exist, and be the target of commit and pr
+    const edges = [
+        { src: '~pr', dest: 'main' },
+        { src: '~commit', dest: 'main' }
+    ];
+
+    // Legacy-mode workflows are always linear, starting with main
+    let src = 'main';
+
+    workflow.forEach((dest) => {
+        edges.push({ src, dest });
+        src = dest;
+    });
+
+    return edges;
+};
+
+/**
+ * Calculate edges of directed graph based on "requires" property of jobs
+ * @method calculateEdges
+ * @param  {Object}       jobs Hash of job configurations
+ * @return {Array}             List of graph edges { src, dest }
+ */
+const calculateEdges = (jobs) => {
+    const edges = [];
+
+    Object.keys(jobs).forEach((j) => {
+        const job = jobs[j];
+        const dest = j;
+
+        if (Array.isArray(job.requires)) {
+            const specialTriggers = job.requires.filter(name => name.charAt(0) === '~');
+            const normalTriggers = job.requires.filter(name => name.charAt(0) !== '~');
+            const isJoin = normalTriggers.length > 1;
+
+            specialTriggers.forEach((src) => {
+                edges.push({ src, dest });
+            });
+
+            normalTriggers.forEach((src) => {
+                const obj = { src, dest };
+
+                if (isJoin) {
+                    obj.join = true;
+                }
+
+                edges.push(obj);
+            });
+        }
+    });
+
+    return edges;
+};
+
+/**
+ * Given a pipeline config, return a directed graph configuration that describes the workflow
+ * @method getWorkflow
+ * @param  {Object}    obj
+ * @param  {Object}    obj.config               A pipeline config
+ * @param  {Object}    obj.config.jobs          Hash of job configs
+ * @param  {Array}     [obj.config.workflow]    Legacy workflow config
+ * @param  {Boolean}   [obj.useLegacy]          Flag to process legacy workflows
+ * @return {Object}                             List of nodes and edges { nodes, edges }
+ */
+const getWorkflow = ({ config: pipelineConfig, useLegacy = false }) => {
+    const jobConfig = pipelineConfig.jobs;
+    let edges = [];
+
+    if (!jobConfig) {
+        throw new Error('No Job config provided');
+    }
+
+    const hasRequiresConfig = Object.keys(jobConfig)
+        .some(j => Array.isArray(jobConfig[j].requires));
+
+    if (useLegacy && !hasRequiresConfig) {
+        // Work out whether there is a user defined workflow,
+        // or if we use the order of jobs defined in jobConfig
+        const workflow = Array.isArray(pipelineConfig.workflow) ?
```

Comment:
When it goes through the config parser, if will add `main` to the workflow.  And workflow saved in the pipeline will always have that main job in workflow.
https://api.screwdriver.cd/v4/pipelines

Can we have some check for `pipelineConfig.workflow`? If it has main then take it out

Reply:
fixed